File: examples/dmd_binary_graycode/cpp/dmd_binary_graycode_example.cxx
#include <ajile/AJObjects.h>
#include <ajile/dmd_constants.h>

#include <vector>
#include <opencv2/opencv.hpp>

#include "example_helper.h"

// helper function which creates a set of binary gray code pattern images
std::vector<cv::Mat> GenerateGrayCodes(int width, int height) {

    // Determine number of required codes and row/column offsets.
    int numColumns = int(ceil(log2(width)));
    int columnShift = int(floor((pow(2.0,numColumns)-width)/2));

    int numRows = int(ceil(log2(height)));
    int rowShift = int(floor((pow(2.0,numRows)-height)/2));

    // Allocate Gray codes.
    std::vector<cv::Mat> grayCodeImages;
    for (int i=0; i<numColumns+numRows+1; i++)
        grayCodeImages.push_back(cv::Mat::zeros(height, width, CV_8UC1));
    
    // Define first code as a white image.
    grayCodeImages[0].setTo(255);

    // Define Gray codes for projector columns.
    for (int c=0; c<width; c++) {
        for (int i=0; i<numColumns; i++) {
            int imageIndex = i+1;
            unsigned char value = 0;
            if ( i > 0 )
                value = (((c+columnShift) >> (numColumns-i-1)) & 1) ^ (((c+columnShift) >> (numColumns-i)) & 1);
            else
                value = (((c+columnShift) >> (numColumns-i-1)) & 1);
            value *= 255;
            cv::rectangle(grayCodeImages[imageIndex], cv::Point(c,0), cv::Point(c,height), value, -1);
        }
    }
            
    // Define Gray codes for projector rows.
    for (int r=0; r<height; r++) {
        for (int i=0; i<numRows; i++) {
            int imageIndex = i+numColumns+1;
            unsigned char value = 0;
            if (i > 0) 
                value = (((r+rowShift) >> (numRows-i-1)) & 1)^(((r+rowShift) >> (numRows-i)) & 1);
            else
                value = (((r+rowShift) >> (numRows-i-1)) & 1);
            value *= 255;
            cv::rectangle(grayCodeImages[imageIndex], cv::Point(0,r), cv::Point(width,r), value, -1);
        }
    }

    return grayCodeImages;
}

// creates an Ajile project and returns in
aj::Project CreateProject(unsigned short sequenceID=1, unsigned int sequenceRepeatCount=0, float frameTime_ms=-1, std::vector<aj::Component> components = std::vector<aj::Component>()) {

    const char* projectName = "dmd_binary_graycode_example";
    if (frameTime_ms < 0)
        frameTime_ms = 100;
    
    // create a new project
    aj::Project project(projectName);
    if (components.size() > 0)
        project.SetComponents(components);

    // generate a list of gray code images (which are opencv matrices)
    std::vector<cv::Mat> grayCodeImages = GenerateGrayCodes(DMD_IMAGE_WIDTH_MAX, DMD_IMAGE_HEIGHT_MAX);
    
    // create the images from the numpy gray code images and add them to our project
    int imageCount = 1;
    for (int i=0; i<grayCodeImages.size(); i++) {
        aj::Image image(imageCount);
        imageCount += 1;
        image.ReadFromMemory((unsigned char*)grayCodeImages[i].data, grayCodeImages[i].rows, grayCodeImages[i].cols, 1, 8, aj::ROW_MAJOR_ORDER, aj::DMD_4500_DEVICE_TYPE);
        project.AddImage(image);
    }

    int numImages = grayCodeImages.size();
    
    // create the sequence
    project.AddSequence(aj::Sequence(sequenceID, projectName, aj::DMD_4500_DEVICE_TYPE, aj::SEQ_TYPE_PRELOAD, sequenceRepeatCount));

    // create a single sequence item, which all the frames will be added to
    project.AddSequenceItem(aj::SequenceItem(sequenceID, 1));

    // create the frames and add them to the project, which adds them to the last sequence item
    for (int i=0; i<numImages; i++) {
        aj::Frame frame;
        frame.SetSequenceID(sequenceID);
        frame.SetImageID(i+1);
        frame.SetFrameTimeMSec(frameTime_ms);
        project.AddFrame(frame);
    }

    return project;
}


int main(int argc, char **argv) {

    return RunExample(&CreateProject, argc, argv);

}

File: examples/dmd_binary_graycode/python/dmd_binary_graycode_example.py
import ajiledriver as aj
import cv2
import numpy
import math

import sys
import os.path
sys.path.insert(0, os.path.split(os.path.realpath(__file__))[0] + "/../../common/python/")
import example_helper

# helper function which creates a set of binary gray code pattern images
def GenerateGrayCodes(width, height):
    # Determine number of required codes and row/column offsets.
    numColumns = int(math.ceil(math.log(width,2)))
    columnShift = int(math.floor((math.pow(2.0,numColumns)-width)/2))

    numRows = int(math.ceil(math.log(height,2)))
    rowShift = int(math.floor((math.pow(2.0,numRows)-height)/2))   

    # Allocate Gray codes.
    grayCodeImages = []
    for i in range(numColumns+numRows+1):
        grayCodeImages.append(numpy.zeros(shape=(height, width, 1), dtype=numpy.uint8))

    # Define first code as a white image.
    grayCodeImages[0].fill(255)

    # Define Gray codes for projector columns.
    for c in range(width):
        for i in range(numColumns):
            imageIndex = i+1
            if i>0:
                grayCodeImages[imageIndex][0, c] = (((c+columnShift) >> (numColumns-i-1)) & 1) ^ (((c+columnShift) >> (numColumns-i)) & 1)
            else: 
                grayCodeImages[imageIndex][0, c] = (((c+columnShift) >> (numColumns-i-1)) & 1)
            grayCodeImages[imageIndex][0, c] *= 255
            cv2.rectangle(grayCodeImages[imageIndex], (c,0), (c,height), int(grayCodeImages[imageIndex][0, c]), -1)

    # Define Gray codes for projector rows.
    for r in range(height):
        for i in range(numRows):
            imageIndex = i+numColumns+1
            if i>0:
                grayCodeImages[imageIndex][r, 0] = (((r+rowShift) >> (numRows-i-1)) & 1)^(((r+rowShift) >> (numRows-i)) & 1)
            else:
                grayCodeImages[imageIndex][r, 0] = (((r+rowShift) >> (numRows-i-1)) & 1)
            grayCodeImages[imageIndex][r, 0] *= 255
            cv2.rectangle(grayCodeImages[imageIndex], (0,r), (width,r), int(grayCodeImages[imageIndex][r, 0]), -1)

    return grayCodeImages


# creates an Ajile project and returns in
def CreateProject(sequenceID=1, sequenceRepeatCount=0, frameTime_ms=-1, components=None):

    projectName = "dmd_binary_graycode_example"
    if frameTime_ms < 0:
        frameTime_ms = 100
    
    # create a new project
    project = aj.Project(projectName)
    if components is not None:
        project.SetComponents(components)

    # generate a list of gray code images (which are numpy matrices)
    grayCodeImages = GenerateGrayCodes(aj.DMD_IMAGE_WIDTH_MAX, aj.DMD_IMAGE_HEIGHT_MAX)
    
    # create the images from the numpy gray code images and add them to our project
    imageCount = 1
    for grayCodeImage in grayCodeImages:
        image = aj.Image(imageCount)
        imageCount += 1
        image.ReadFromMemory(grayCodeImage, 8, aj.ROW_MAJOR_ORDER, aj.DMD_4500_DEVICE_TYPE)
        project.AddImage(image)
        # add preview images (which will be used by the GUI only for display, ignore if not opening the example in the GUI)
        example_helper.AddPreviewImage(project, grayCodeImage, image.ID(), image.ID(), "graycode_" + str(image.ID()), 1)

    numImages = len(grayCodeImages)
    
    # create the sequence
    project.AddSequence(aj.Sequence(sequenceID, projectName, aj.DMD_4500_DEVICE_TYPE, aj.SEQ_TYPE_PRELOAD, sequenceRepeatCount))

    # create a single sequence item, which all the frames will be added to
    project.AddSequenceItem(aj.SequenceItem(sequenceID, 1))

    # create the frames and add them to the project, which adds them to the last sequence item
    for i in range(numImages):
        frame = aj.Frame()
        frame.SetSequenceID(sequenceID)
        frame.SetImageID(i+1)
        frame.SetFrameTimeMSec(frameTime_ms)
        project.AddFrame(frame)

    return project

        
if __name__ == "__main__":

    example_helper.RunExample(CreateProject)
